src/metadata/authorization_server.rs

```diff
@@ -0,0 +1,155 @@
+    pushed_authorization_request_endpoint: Option<ParUrl>,
+    #[serde(default)]
+    require_pushed_authorization_requests: bool,
+    additional_fields: Map<String, Json>,
+}
+
+impl AuthorizationServerMetadata {
+    #[cfg(test)]
+    pub fn new(issuer: IssuerUrl, token_endpoint: TokenUrl) -> Self {
+        Self {
+            issuer,
+            authorization_endpoint: Default::default(),
+            token_endpoint,
+            jwks_uri: Default::default(),
+            registration_endpoint: Default::default(),
+            scopes_supported: Default::default(),
+            response_types_supported: Default::default(),
+            response_modes_supported: Default::default(),
+            grant_types_supported: Default::default(),
+            revocation_endpoint: Default::default(),
+            introspection_endpoint: Default::default(),
+            code_challenge_methods_supported: Default::default(),
+            pre_authorized_grant_anonymous_access_supported: false,
+            pushed_authorization_request_endpoint: Default::default(),
+            require_pushed_authorization_requests: Default::default(),
+            additional_fields: Default::default(),
+        }
+    }
+
+    field_getters_setters![
+        pub self [self] ["authorization server metadata value"] {
+            set_issuer -> issuer[IssuerUrl],
+            set_authorization_endpoint -> authorization_endpoint[Option<AuthUrl>],
+            set_token_endpoint -> token_endpoint[TokenUrl],
+            set_jwks_uri -> jwks_uri[Option<JsonWebKeySetUrl>],
+            set_registration_endpoint -> registration_endpoint[Option<RegistrationUrl>],
+            set_scopes_supported -> scopes_supported[Option<Vec<Scope>>],
+            set_response_types_supported -> response_types_supported[Option<Vec<ResponseType>>],
+            set_response_modes_supported -> response_modes_supported[ResponseModes],
+            set_grant_types_supported -> grant_types_supported[GrantTypesSupported],
+            set_revocation_endpoint -> revocation_endpoint[Option<RevocationUrl>],
+            set_introspection_endpoint -> introspection_endpoint[Option<IntrospectionUrl>],
+            set_code_challenge_methods_supported -> code_challenge_methods_supported[Option<Vec<PkceCodeChallengeMethod>>],
+            set_pre_authorized_grant_anonymous_access_supported -> pre_authorized_grant_anonymous_access_supported[bool],
+            set_pushed_authorization_request_endpoint -> pushed_authorization_request_endpoint[Option<ParUrl>],
+            set_require_pushed_authorization_requests -> require_pushed_authorization_requests[bool],
+        }
+    ];
+
+    pub fn additional_fields(&self) -> &Map<String, Json> {
+        &self.additional_fields
+    }
+
+    pub fn additional_fields_mut(&mut self) -> &mut Map<String, Json> {
+        &mut self.additional_fields
+    }
+}
+
+impl MetadataDiscovery for AuthorizationServerMetadata {
```

Comment:
I don't know if it will be possible to use a trait for the validation, considering the information needed will vary (e.g. the validation of an authorization server metadata depends on credential issuer metadata https://openid.net/specs/openid-4-verifiable-credential-issuance-1_0-ID1.html#section-11.2.3-2.2) -- but I know that this will be implemented later